core/retry.go

```diff
@@ -0,0 +1,130 @@
+package retry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+/*
+This retry library was inspired by and uses Cenk Alti (https://github.com/cenkalti) backoff library (https://github.com/cenkalti/backoff).
+We would like to thank him for his great work.
+*/
+
+/*
+Note we use a custom Permanent error type for asserting Permanent Erros within the retry library.
+We do not implement an explicit Transient error type and operate under the assumption that all errors that are not Permanent are Transient.
+*/
+type PermanentError struct {
+	Inner error
+}
+
+func (e PermanentError) Error() string { return e.Inner.Error() }
+func (e PermanentError) Unwrap() error {
+	return e.Inner
+}
+func (e PermanentError) Is(err error) bool {
+	_, ok := err.(PermanentError)
+	return ok
+}
+
+const (
+	MinDelay               = 1 * time.Second
+	MaxInterval            = 60 * time.Second
+	MaxElapsedTime         = 0 * time.Second
+	RetryFactor    float64 = 2
+	NumRetries     uint64  = 3
+)
+
+// Same as Retry only that the functionToRetry can return a value upon correct execution
+func RetryWithData[T any](functionToRetry func() (T, error), minDelay time.Duration, factor float64, maxTries uint64, maxInterval time.Duration, maxElapsedTime time.Duration) (T, error) {
```

Comment:
These functions receive a lot of parameters and it is easy to get lost. I would refactor this to receive a struct instead with all the config. This will also make passing a default (which is what we are doing in most if not all retryable functions) much easier, as you would only pass one param and reuse the struct. I'll give you a hint:

```go
type RetryConfig {
MinDelay time.Duration
MaxInterval time.Duration
MaxDelay time.Duration
RetryFactor float64
NumRetries uint64
}

func RetryWithData[T any](functionToRetry func() (T,error), config *RetryConfig) { }

// The you would you use like so 
func foo() {
RetryWithData(funcToRetry, &RetryConfig{ RetryFactor: 5 , ...})
}
```